cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -459,16 +465,199 @@ class self_comparator {
    * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
    */
   template <typename Nullate>
-  less_comparator<Nullate> device_comparator(Nullate nullate = {}) const
+  less_comparator<device_row_comparator, Nullate> device_comparator(Nullate nullate = {}) const
   {
-    return less_comparator<Nullate>{device_row_comparator<Nullate>(
+    return less_comparator<device_row_comparator, Nullate>{device_row_comparator<Nullate>(
       nullate, *d_t, *d_t, d_t->depths(), d_t->column_order(), d_t->null_precedence())};
   }
 
  private:
   std::shared_ptr<preprocessed_table> d_t;
 };
 
+enum class lhs_index_type : size_type {};
+enum class rhs_index_type : size_type {};
+
+struct make_lhs_index {
+  __device__ lhs_index_type operator()(size_type i) const { return static_cast<lhs_index_type>(i); }
+};
+
+struct make_rhs_index {
+  __device__ rhs_index_type operator()(size_type i) const { return static_cast<rhs_index_type>(i); }
+};
```

Comment:
We shouldn't need `make_` factories in C++17. Especially so here because there's no template types. 

I'd suggest just defining a simple, custom iterator. 

```suggestion
template <typename T>
struct strong_index_iterator{
  explicit strong_index_iterator(T begin = T{0}){};
 // iterator boilerplate
};
using lhs_iterator = strong_index_iterator<lhs_index_t>;
using rhs_iterator = strong_index_iterator<rhs_index_t>;
```

Example:
```
thrust::foo(make_lhs_index_counting_iterator(0), make_lhs_index_counting_iterator(N))

// vs
thrust::foo(lhs_iterator{0}, lhs_iterator{N});
```

Reply:
I only added `make_lhs_index` / `make_rhs_index` because of some issue with passing a lambda to `cudf::make_counting_transform_iterator`. I'll investigate the solution you proposed.

Reply:
This is now resolved. It's significantly more lines of code but I'm not sure what simplifications would be appropriate, if any.

Reply:
I should have mentioned that you can use `thrust::iterator_adaptor` or `thrust::iterator_facade` to reduce most of the boilerplate.

https://thrust.github.io/doc/classthrust_1_1iterator__adaptor.html
https://thrust.github.io/doc/classthrust_1_1iterator__facade.html

I spent a few minutes on it but couldn't quite get it to work. I'll let you pick up where I left off: https://godbolt.org/z/Kqqf6KKGa 

The `boost::iterator_facade` docs are _far_ better than Thrust's: https://www.boost.org/doc/libs/1_65_0/libs/iterator/doc/iterator_facade.html#tutorial-example

Reply:
That's a great idea. I played with this a little bit more: https://godbolt.org/z/GvTK5TaE9

This compiles but I haven't tried to simplify or clean it up yet. I'll pick it up here on Monday.

Reply:
Adapted in 1834095dfc54dc0941d219353f21e2ad8d3d303c. Turned out to be a very simple copy-paste from the Godbolt implementation I made. 💯

Reply:
What is your original issues with using `thrust::make_transform_iterator`?

Reply:
I think the problem was with computing distances / difference types. A custom iterator class solves this.